ddl/column.go

```diff
@@ -643,17 +655,354 @@ func (w *worker) doModifyColumn(
 		}
 	})
 
-	if newAutoRandBits > 0 {
-		if err := checkAndApplyNewAutoRandomBits(job, t, tblInfo, newCol, oldName, newAutoRandBits); err != nil {
+	if updatedAutoRandomBits > 0 {
+		if err := checkAndApplyNewAutoRandomBits(job, t, tblInfo, newCol, oldColName, updatedAutoRandomBits); err != nil {
+			return ver, errors.Trace(err)
+		}
+	}
+
+	if !needChangeColumnData(oldCol, newCol) {
+		return w.doModifyColumn(t, job, dbInfo, tblInfo, newCol, oldCol, pos)
+	}
+
+	if changingCol == nil {
+		changingColPos := &ast.ColumnPosition{Tp: ast.ColumnPositionNone}
+		newColName := model.NewCIStr(fmt.Sprintf("%s%s", changingColumnPrefix, newCol.Name.O))
+		if mysql.HasPriKeyFlag(oldCol.Flag) {
+			job.State = model.JobStateCancelled
+			msg := "enableChangeColType is true and this column has primary key flag"
+			return ver, errUnsupportedModifyColumn.GenWithStackByArgs(msg)
+		}
+		// TODO: Check whether we need to check OriginDefaultValue.
+		changingCol = newCol.Clone()
+		changingCol.Name = newColName
+		changingCol.ChangeStateInfo = &model.ChangeStateInfo{DependencyColumnOffset: oldCol.Offset}
+		_, _, _, err := createColumnInfo(tblInfo, changingCol, changingColPos)
+		if err != nil {
+			job.State = model.JobStateCancelled
+			return ver, errors.Trace(err)
+		}
+
+		idxInfos, offsets := findIndexesByColName(tblInfo.Indices, oldCol.Name.L)
+		changingIdxs = make([]*model.IndexInfo, 0, len(idxInfos))
+		for i, idxInfo := range idxInfos {
+			newIdxInfo := idxInfo.Clone()
+			newIdxInfo.Name = model.NewCIStr(fmt.Sprintf("%s%s", changingIndexPrefix, newIdxInfo.Name.O))
+			newIdxInfo.ID = allocateIndexID(tblInfo)
+			newIdxInfo.Columns[offsets[i]].Name = newColName
+			newIdxInfo.Columns[offsets[i]].Offset = newCol.Offset
```

Comment:
Should use `changingCol.Offset`? since `newCol.Offset` actually equal to `oldCol.Offset`.

Reply:
Here is an example:

I just mock a sleep in code:

![image](https://user-images.githubusercontent.com/26020263/90021604-14edbb80-dce4-11ea-8dc2-61420ac8caa6.png)

1. execute in session 1:
```sql
create table t (a bigint, b int, index idx(a));
alter table t modify a int unsigned;
```

2. execute in session 2 and the DDL job state in `StateWriteReorganization`:

```sql
test> show create table t;
+-------+-------------------------------------------------------------+
| Table | Create Table                                                |
+-------+-------------------------------------------------------------+
| t     | CREATE TABLE `t` (                                          |
|       |   `a` bigint(20) DEFAULT NULL,                              |
|       |   `b` int(11) DEFAULT NULL,                                 |
|       |   KEY `idx` (`a`)                                           |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+-------------------------------------------------------------+
1 row in set
Time: 0.011s
test> set sql_mode='';
Query OK, 0 rows affected
Time: 0.001s
test> insert into t values (9223372036854775807,1);
Query OK, 1 row affected
Time: 0.002s
test> select * from t;
+---------------------+---+
| a                   | b |
+---------------------+---+
| 9223372036854775807 | 1 |
+---------------------+---+
```

3.  After DDL job finish:

```sql
test> show create table t;
+-------+-------------------------------------------------------------+
| Table | Create Table                                                |
+-------+-------------------------------------------------------------+
| t     | CREATE TABLE `t` (                                          |
|       |   `a` int(10) unsigned DEFAULT NULL,                        |
|       |   `b` int(11) DEFAULT NULL,                                 |
|       |   KEY `idx` (`a`)                                           |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+-------------------------------------------------------------+
1 row in set
Time: 0.011s
test> admin check table t;
(1105, 'col a, handle 2, index:types.Datum{k:0x2, decimal:0x0, length:0x0, i:9223372036854775807, collation:"", b:[]uint8(nil), x:interface {}(nil)} != record:types.Datum{k:0x2, decimal:0x0, length:0x0, i:4294967295, collation:"", b:[]uint8(nil), x:interface {}(nil)}')
test> select * from t;
+------------+---+
| a          | b |
+------------+---+
| 4294967295 | 1 |
+------------+---+
```

Reply:
I guess this is because some truncation is not handled yet?

Reply:
@crazycs520 @bb7133 
In the above case, we can insert overflows value, because we don't handle the empty sql_mode(I added TODO in code) in DML. If the sql_mode is the default, executing the above insert statement will return `overflows` error in `WriteReorg`.

Reply:
> I guess this is because some truncation is not handled yet?

This is cause by the offset of the index column is wrong, I was mentioned in my first comment.

> Should use `changingCol.Offset`? since `newCol.Offset` actually equal to `oldCol.Offset`.